wire-runtime/src/main/java/com/squareup/wire/ProtoAdapter.java

```diff
@@ -485,4 +492,84 @@ public String toString(E value) {
       this.value = value;
     }
   }
+
+  private static final class MapProtoAdapter<K, V> extends ProtoAdapter<Map<K, V>> {
+    private final MapEntryProtoAdapter<K, V> entryAdapter;
+
+    MapProtoAdapter(ProtoAdapter<K> keyAdapter, ProtoAdapter<V> valueAdapter) {
+      super(FieldEncoding.LENGTH_DELIMITED, null);
+      entryAdapter = new MapEntryProtoAdapter<>(keyAdapter, valueAdapter);
+    }
+
+    @Override public int encodedSize(Map<K, V> value) {
+      throw new UnsupportedOperationException("Repeated values can only be sized with a tag.");
+    }
+
+    @Override public int encodedSizeWithTag(int tag, Map<K, V> value) {
+      int size = 0;
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        size += entryAdapter.encodedSizeWithTag(tag, entry);
+      }
+      return size;
+    }
+
+    @Override public void encode(ProtoWriter writer, Map<K, V> value) {
+      throw new UnsupportedOperationException("Repeated values can only be encoded with a tag.");
+    }
+
+    @Override public void encodeWithTag(ProtoWriter writer, int tag, Map<K, V> value)
+        throws IOException {
+      for (Map.Entry<K, V> entry : value.entrySet()) {
+        entryAdapter.encodeWithTag(writer, tag, entry);
+      }
+    }
+
+    @Override public Map<K, V> decode(ProtoReader reader) throws IOException {
+      K key = null;
+      V value = null;
+
+      long token = reader.beginMessage();
+      for (int tag; (tag = reader.nextTag()) != -1;) {
+        switch (tag) {
+          case 1: key = entryAdapter.keyAdapter.decode(reader); break;
+          case 2: value = entryAdapter.valueAdapter.decode(reader); break;
+          default: // Ignore unknown tags in map entries.
+        }
+      }
+      reader.endMessage(token);
+
+      if (key == null) throw new IllegalStateException("Map entry with null key");
+      if (value == null) throw new IllegalStateException("Map entry with null value");
+      return Collections.singletonMap(key, value);
```

Comment:
This is because of how the encoding works.

Think of how we do repeated messages: we create an empty list and append a single message at a time as we see them. This is because the proto adapter only knows how to decode the message but the parent adapter is the one that matches on the tag and delegates.

Maps are effectively a repeated message of key/value pairs. Unlike lists, however, there are two types that we need to decode along with the enclosing message. This means that we either can re-use the `Map<K, V>` adapter which is needed for encoding and getting the encoded size for reading single-element `Map<K, V>` instances, or we can create a separate adapter for just `Map.Entry<K, V>`.

If you compare the implementation of `Collections.singletonMap` with `SimpleMapEntry` they're effectively the same class. The difference is that `Map` has a `putAll(Map)` method where there is no `put(Map.Entry)`. This means that if we want the calling code to be a one-liner, the singleton map + `putAll` is the easiest way to go.

If we use an adapter for `Map.Entry` instead, you also have a problem of encoding. If you call `asRepeated()` which returns a `ProtoAdapter<List<Map.Entry<K, V>>>` we have no way to represent the list as that type without a copy of `.entrySet()`.

Simply put: neither `ProtoAdapter<Map<K, V>>` nor `ProtoAdapter<Map.Entry<K, V>>` ends up being ergonomic for all three methods, but I think `Map` is more ergonomic (and there's parity with `List`).